traits/ctraits.c

```diff
@@ -2908,22 +2897,31 @@ static trait_setattr setattr_handlers[] = {
     /*  End of __setstate__ method entries */
     NULL};
 
-static int
-trait_init(trait_object *trait, PyObject *args, PyObject *kwds)
+
+trait_object *
+trait_new (PyTypeObject *trait_type, PyObject *args, PyObject *kw)
 {
-    int kind;
+    int kind, n;
+    trait_object *trait;
 
-    if (!PyArg_ParseTuple(args, "i", &kind)) {
-        return -1;
+    n = PyTuple_GET_SIZE(args);
+
+    if (n == 0){
+        kind = 0;
+    }
+    else if (!PyArg_ParseTuple(args, "i", &kind)) {
```

Comment:
I think we should be raising here if there are keyword arguments provided (the current behaviour is that those arguments are silently ignored). Can we add a block above that checks `kw`, along the lines of (untested):
```
    if (kw != NULL && PyDict_GET_SIZE(kw) != 0) {
        PyErr_SetString(PyExc_TypeError, "CTrait takes no keyword arguments");
        return NULL;
    }
```
... along with a unit test that exercises this block.